scripts/externalTests/common.sh

```diff
@@ -106,25 +106,31 @@ function setup_solc
 
 function download_project
 {
-    local repo="$1"
+    local repo_url="$1"
     local ref_type="$2"
-    local solcjs_ref="$3"
+    local ref="$3"
     local test_dir="$4"
 
-    [[ $ref_type == commit || $ref_type == branch || $ref_type == tag ]] || assertFail
+    [[ $ref_type == commit || $ref_type == branch || $ref_type == tag || $ref_type == release ]] || assertFail
 
-    printLog "Cloning ${ref_type} ${solcjs_ref} of ${repo}..."
-    if [[ $ref_type == commit ]]; then
-        mkdir ext
-        cd ext
-        git init
-        git remote add origin "$repo"
-        git fetch --depth 1 origin "$solcjs_ref"
-        git reset --hard FETCH_HEAD
-    else
-        git clone --depth 1 "$repo" -b "$solcjs_ref" "$test_dir/ext"
-        cd ext
+    printLog "Cloning ${repo_url}..."
+    # Clone the repo ignoring all blobs until needed by git.
+    # This allows access to commit history but with a fast initial clone
+    git clone --filter=blob:none "$repo_url" "$test_dir/ext"
+    cd "$test_dir/ext"
+
+    # If the ref_type is 'release' try to use the latest tag as ref
+    # NOTE: Sadly this will not work with monorepos and may not always
+    # return the latest tag.
+    if [[ $ref_type == release ]]; then
+       ref=$(git tag --sort=-v:refname | head --lines=1)
     fi
```

Comment:
But I'm assuming that it at least gets it right for the tests we have?

You could also make it look only at annotated tags. These are usually meant for releases because they have a date and description. Though I'm not sure if everyone sticks to that in practice.

Reply:
> You could also make it look only at annotated tags. These are usually meant for releases because they have a date and description. Though I'm not sure if everyone sticks to that in practice.

Most of them don't do that unfortunately.

Reply:
Honestly, not sure if it is worth the amount of work to workaround these stuffs.

Reply:
> The more complex the logic for this is, the more likely it'll break... if it's too complicated to reliably determine the latest release, we _can_ also just hardcode it and use traditional methods to remind us to just regularly check/bump them... but if the current method works for the tests we have, I'd just stick with it for now. But yeah, when suggesting to track the respective latest release automatically, I didn't think through the fact that that may in fact be quite complicated to do reliably in full generality - no need to go miles out of our way for it, I'd say. _Or_ we could just define a "fetch latest release" mechanism in the respective script for the particular external test - which can then either use the simplest mechanism to fetch it (if any) for each external repo - or just a fixed one (if none). Really no need to try to solve the problem for every possible repo imaginable, is there :-)?

I completely agree. I tried to find a command that could at least provide an intersection of the latest tags across our external tests only.


> Actually, I'd myself personally even remove the `download_project` function entirely from `common.sh` and just do it manually in each of the external test script themselves, that looks to me like it'd be the much simpler choice and the amount of common code in this is quite negligible :-) - but I don't want to weigh in heavily on this from the side, you're the ones taking care of these scripts, so 100% your call :-).

Yes, I mean, it's already done, so let's leave it as it is for now. We'll eventually migrate everything to Python anyway, and I believe the maintenance burden will be much less.

Reply:
ok, if you think it adds too much complexity let's forget about blacklisting tags and keep Chainlink at `master`. If we did add it, I'd still prefer to hide that in `download_project` though - the idea was that most of the things in the project-specific scripts are just invocations of reusable functionality. You can mix and match any steps from the toolbox you need, while still keeping the recipe concise. Unfortunately with this being written in Bash, it sill ends up being quite messy, but since we'll eventually rewrite it in Python, all the more reason not to change the whole design right now.

Also, I just noticed that blacklisting tags for Chainlink would be pointless after all, because we use our fork from `solidity-external-tests` and not the upstream. So problem solved :)